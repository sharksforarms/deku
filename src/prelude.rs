pub use crate::{error::DekuError, BitsReader, BitsWriter, DekuRead, DekuWrite};
pub use bitvec::{
    order::BitOrder, order::Lsb0, order::Msb0, slice::AsBits, slice::BitSlice, vec::BitVec,
};
