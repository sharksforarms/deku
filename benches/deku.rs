use criterion::{black_box, criterion_group, criterion_main, Criterion};
use deku::prelude::*;

#[derive(Debug, PartialEq, DekuRead)]
struct DekuByte {
    data: u8
}

#[derive(Debug, PartialEq, DekuRead)]
#[deku(id_type = "u8")]
enum DekuEnum {
    #[deku(id = "0x01")]
    VariantA(u8),
}

#[derive(Debug, PartialEq, DekuRead)]
struct DekuVec {
    count: u8,
    #[deku(len = "count")]
    data: Vec<u8>
}

fn deku_read_byte(input: &[u8]) {
    let (_rest, v) = DekuByte::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuByte { data: 0x01 }, v);
}

fn deku_read_enum(input: &[u8]) {
    let (_rest, v) = DekuEnum::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuEnum::VariantA(0x02), v);
}

fn deku_read_vec(input: &[u8]) {
    let (_rest, v) = DekuVec::from_bytes((input, 0)).unwrap();
    assert_eq!(100, v.count);
    assert_eq!(100, v.data.len());
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("deku_read_byte", |b| {
        b.iter(|| deku_read_byte(black_box([0x01].as_ref())))
    });

    c.bench_function("deku_read_enum", |b| {
        b.iter(|| deku_read_enum(black_box([0x01, 0x02].as_ref())))
    });

    let deku_read_vec_input = {
        let mut v = [0xFFu8; 101].to_vec();
        v[0] = 100u8;
        v
    };
    c.bench_function("deku_read_vec", |b| {
        b.iter(|| deku_read_vec(black_box(&deku_read_vec_input)))
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
