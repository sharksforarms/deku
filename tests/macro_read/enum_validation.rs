use deku::prelude::*;

// test required attributes
#[derive(DekuRead)]
enum Test1 {}

// test conflict `id_type` and `id`
#[derive(DekuRead)]
#[deku(id_type = "u8", id = "test")]
enum Test2 {}

// test conflict `id` and `id_pat`
#[derive(DekuRead)]
#[deku(id_type = "u8")]
enum Test3 {
    #[deku(id = "1", id_pat = "2..=3")] A(u8),
}

// test `id_type` only allowed on enum
#[derive(DekuRead)]
#[deku(id_type = "u8")]
struct Test4 {
    a: u8
}

// test `bits` only allowed on enum
#[derive(DekuRead)]
#[deku(bits = "1")]
struct Test5 {
    a: u8
}

// test `id_bytes` only allowed on enum
#[derive(DekuRead)]
#[deku(bits = "1")]
struct Test6 {
    a: u8
}

// test `id` only allowed on enum
#[derive(DekuRead)]
#[deku(id = "test")]
struct Test7 {
    a: u8
}

// test `bits` cannot be used with `id`
#[derive(DekuRead)]
#[deku(id = "test", bits = "4")]
enum Test8 {
    A
}

// test `id_bytes` cannot be used with `id`
#[derive(DekuRead)]
#[deku(id = "test", id_bytes = "4")]
enum Test9 {
    A
}

fn main() {}
