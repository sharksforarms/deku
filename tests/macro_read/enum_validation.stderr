error: `id_type` or `id` must be specified on enum
 --> $DIR/enum_validation.rs:5:6
  |
5 | enum Test1 {}
  |      ^^^^^

error: conflicting: both `id_type` and `id` specified on enum
  --> $DIR/enum_validation.rs:10:6
   |
10 | enum Test2 {}
   |      ^^^^^

error: conflicting: both `id` and `id_pat` specified on variant
  --> $DIR/enum_validation.rs:16:17
   |
16 |     #[deku(id = "1", id_pat = "2..=3")] A(u8),
   |                 ^^^

error: `id_type` only supported on enum
  --> $DIR/enum_validation.rs:21:18
   |
21 | #[deku(id_type = "u8")]
   |                  ^^^^

error: `id_bits` only supported on enum
  --> $DIR/enum_validation.rs:27:10
   |
27 | #[derive(DekuRead)]
   |          ^^^^^^^^
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: `id_bits` only supported on enum
  --> $DIR/enum_validation.rs:34:10
   |
34 | #[derive(DekuRead)]
   |          ^^^^^^^^
   |
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: `id` only supported on enum
  --> $DIR/enum_validation.rs:42:13
   |
42 | #[deku(id = "test")]
   |             ^^^^^^
